### DISCUSSION TITLE
How can i push a file video to livestream.

### DISCUSSION THREAD

**Discussion by DrDepTraii** [HaishinKit/HaishinKit.swift#1416] [Q&A]
I've looked through the examples but still haven't found a solution. Please provide me the code to solve the problem.

**Comment by DrDepTraii** [comment]
import UIKit
import AVFoundation
import HaishinKit
import VideoToolbox

class ViewController: UIViewController {

    private lazy var rtmpConnection: RTMPConnection = {
            let conneciton = RTMPConnection()
           
            return conneciton
        }()
    var rtmpStream: RTMPStream!
    var videoURL: URL!
    var videoData: Data?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Khởi tạo RTMPConnection
        rtmpConnection = RTMPConnection()
        rtmpConnection.addEventListener(.rtmpStatus, selector: #selector(rtmpStatusEvent), observer: self)
        rtmpConnection.addEventListener(.ioError, selector: #selector(rtmpErrorHandler), observer: self)
        
        // Cấu hình RTMPStream
        rtmpStream = RTMPStream(connection: rtmpConnection)
        // Cấu hình RTMPStream
        rtmpStream.videoSettings.videoSize = .init(width: 1280, height: 720)
        rtmpStream.videoSettings.profileLevel = kVTProfileLevel_H264_Baseline_AutoLevel as String

        
        // Đường dẫn đến tệp video MOV
        guard let videoUrl = Bundle.main.url(forResource: "IMG_0255", withExtension: "MOV") else {
                   print("no video")
                   return
               }
        self.videoURL = videoUrl
        
        
        // Tải dữ liệu video
        do {
            videoData = try Data(contentsOf: videoURL)
        } catch {
            print("Failed to load video data: \(error)")
        }
        
        // Kết nối và phát trực tiếp
        rtmpConnection.connect("rtmp://a.rtmp.youtube.com/live2") //URL của máy chủ RTMP
    }
    
    @objc private func rtmpErrorHandler(_ notification: Notification) {
                //rtmpConnection.connect(Preference.defaultInstance.uri!)
            }

        @objc private func rtmpStatusEvent(_ status: Notification) {
                let e = Event.from(status)
                guard
                    let data: ASObject = e.data as? ASObject,
                    let code: String = data["code"] as? String else {
                    return
                }
                switch code {
                case RTMPConnection.Code.connectSuccess.rawValue:
                    startStreaming()
                    rtmpStream.publish("my_key")
                    
                default:
                    break
                }
            }
   
    func startStreaming() {
        // Kiểm tra xem đã có dữ liệu video và RTMPStream đã sẵn sàng
        guard let videoData = videoData, rtmpStream != nil else { return }
        
        // Phát trực tiếp video
        let videoSampleBuffer = CMSampleBufferCreateReadyWithNSData(allocator: nil, data: videoData as NSData, formatDescription: nil, sampleTiming: nil, sampleBuffer: nil)
        guard let a = videoSampleBuffer else {return}
        rtmpStream.append(a)
    }
    
    func stopStreaming() {
        // Ngừng phát trực tiếp và ngắt kết nối
        rtmpStream.close()
        rtmpConnection.close()
    }
    
    // Hàm hỗ trợ tạo CMSampleBuffer từ NSData
    func CMSampleBufferCreateReadyWithNSData(allocator: CFAllocator?, data: NSData, formatDescription: CMFormatDescription?, sampleTiming: CMSampleTimingInfo?, sampleBuffer: CMSampleBuffer?) -> CMSampleBuffer? {
        let dataLength = data.length
        let bytes = UnsafeMutablePointer<Int8>.allocate(capacity: dataLength)
        data.getBytes(bytes, length: dataLength)
        
        var blockBuffer: CMBlockBuffer?
        var status = CMBlockBufferCreateWithMemoryBlock(allocator: kCFAllocatorDefault, memoryBlock: bytes, blockLength: dataLength, blockAllocator: kCFAllocatorNull, customBlockSource: nil, offsetToData: 0, dataLength: dataLength, flags: 0, blockBufferOut: &blockBuffer)
        
        if status != kCMBlockBufferNoErr {
            return nil
        }
        
        let sampleSizeArray = [dataLength]
        status = CMBlockBufferReplaceDataBytes(with: bytes, blockBuffer: blockBuffer!, offsetIntoDestination: 0, dataLength: dataLength)
        
        if status != kCMBlockBufferNoErr {
            return nil
        }
        
        var sampleBuffer: CMSampleBuffer?
        var sampleSizes = sampleSizeArray
        status = CMSampleBufferCreateReady(allocator: kCFAllocatorDefault, dataBuffer: blockBuffer!, formatDescription: formatDescription, sampleCount: 1, sampleTimingEntryCount: 0, sampleTimingArray: nil, sampleSizeEntryCount: 1, sampleSizeArray: &sampleSizes, sampleBufferOut: &sampleBuffer)
        
        if status != kCMBlockBufferNoErr {
            return nil
        }
        
        return sampleBuffer
    }
}


this is my code. 
I don't understand why my live stream isn't coming up

**Comment by shogo4405** [comment]
Close as https://github.com/shogo4405/HaishinKit.swift/discussions/1358
